### ISSUE TITLE
Functional pattern for retrieving deeply nested data

### ISSUE THREAD

**Issue by bunkat** [optimizely/nuclear-js#33]
What is the correct way to load deeply nested data using NuclearJS?  The chat example shows a basic parent/child relationship where the children are wholly contained within the parent, but in my case the parent and child are related via references.  Given that, I'm not sure what the most pragmatic way to load additional levels of data would be (especially since these deeper levels may still need to be fetched from the server if they aren't in a store).

Here is an example of the data in my store:

````
{
    tasks: {
        1:  {id: 1, name: 'task1', contacts: [{id: 1}], files: [ {id: 1}, {id: 2}]},
        2:  {id: 2, name: 'task2', files: [ {id: 1}, {id: 4}]},
    },
    files: {
        1:  {id: 1, name: 'file1', contacts: [ {id: 1}, {id: 2}]},
        2:  {id: 2, name: 'file2', contacts: [ {id: 1}]},
    },
   contacts: {
        1:  {id: 1, name: 'contact1'},
    }
}      
````

Currently I have a `current-task-id-store` that tracks the currently selected task.  I can then use this state to grab the files for the task.

````
// current task id store
function setCurrentTaskID(state, { taskID}) {
  // return the new value of the store's state
  return taskID
}

// task getters
exports.currentTask = [
  ['currentTaskID'],
  exports.tasksMap,
  (currentTaskID, tasksMap) => tasksMap.get(currentTaskID)
]

exports.currentTaskFiles = [
  exports.currentTask,
  ['filesMap']
  (currentTask, filesMap) => ...get files from store / dispatch fetch from server...
]
````

I then use `getDataBindings` to get the set of files that I need to render.

````
// fileSection component
var FileSection = React.createClass({
  mixins: [reactor.ReactMixin],

  getDataBindings() {
    return {
      files: Task.getters.currentTaskFiles
    }
  }, ...

```

However, now I need to load the `contact` references for each of those files so that I can render them.  I'm not sure how to write a functional getter that would load/fetch them. I thought about just using a binding to get the `contactsMap` and looking up the values directly but this doesn't seem right and doesn't work in cases where the contact needs to be fetched from the server.

**Comment by jordangarcia** [comment]
Hi bunkat,

You are definitely on the right track by splitting your data up in to separate stores then reconstructing the relationships in getters.

This is a common question with NuclearJS and Flux in general, and that is how to load / fetch data into your system.

My opinion from working with Flux & NuclearJS in large codebases is to put all fetch or "load" operations within actions and ideally have them be idempotent, meaning if they are called again redundantly then it no-ops and defers to the version in memory in a store.

The question then comes to when to do initiate the fetching.  At Optimizely we use a router which chains functions together to fetch data, set state then render the page.  The nice thing about that pattern is each function can block on some async action.  For example we have a `fetchProject` action which does an API call to get the project and blocks all other execution until that is done, that way when our components are rendered we always assume a valid project is loaded in our flux stores.

tl;dr; I would put your asnyc operations in actions and find a good place to invoke them whether it be in your initialization script or within a `componentWillMount` lifecycle.

**Comment by bunkat** [comment]
Thanks Jordan, appreciate the help.  So do you end up calling actions from your getters?  For example, in my example about getting files for  task:

````
exports.currentTaskFiles = [
  exports.currentTask,
  ['filesMap']
  (currentTask, filesMap) => ...get files from store / dispatch fetch from server...
]
````

I want to be able to return whatever I have from the store, but then I also need to be able to fetch data from the server if any of the referenced files haven't been loaded yet.

So is this roughly the flow of data?

* Component calls getter to retrieve files
* Getter returns files that have been previously fetched
* Getter calls action to load additional references
* Component renders with initial data
* Action makes async call, dispatches 'FETCH_COMPLETE' message when complete
* ApiStore listens for 'FETCH_COMPLETE' and adds results to store
* Component observes change and re-calls getter  (NuclearJS handles this?)
* Component re-renders with updated data (NuclearJS handles this?)

And then for the contact rendering, it will call a `getContactById(contactid)` action in `componentWillMount` to retrieve the nested data to render (instead of calling a functional getter).

If that's the case, then I'm wondering why have a getter for `currentTaskFiles` at all.  Should this just be an action that returns data from the store and then makes the async call if needed?  It seems like almost every call will have this pattern (check store, async fetch missing data) unless you load everything up front, but even then you still need to refresh existing data from the server.

**Comment by jordangarcia** [comment]
Hi bunkat,

>>  So do you end up calling actions from your getters? For example, in my example about getting files for task:

No, not quite.  Getters should be pure functions that simply transform and compose data in your App State.  Actions start data fetching should be called in your app initialization or in the component lifecycle methods, such as `componentWillMount`.  

In your example if you have a `ContactComponent` you could put the `fetchCnontactById(contactId)` action call within the `componentWillMount` or as a click  handler for the parent component where you select the contact from within the UI.

I am working on a async example this weekend that should clarify some of these patterns.

**Comment by jtremback** [comment]
Is the async example done yet?

**Comment by jordangarcia** [comment]
its close to done,  in the meantime you can checkout
https://github.com/optimizely/nuclear-js/tree/rest-api-example
